features/steps/validation_handling.py

```diff
@@ -116,16 +117,33 @@ def handle_then(context, fn, **kwargs):
     # if 'instances' are not actual ifcopenshell.entity_instance objects, but e.g. tuple of string values then get the actual instances from the stack tree
     activation_instances = misc.do_try(lambda: get_stack_tree(context)[-1], instances)
 
-    validation_outcome = ValidationOutcome(
-        outcome_code=ValidationOutcomeCode.EXECUTED,  # "Executed", but not no error/pass/warning #deactivated for now
-        observed=None,
-        expected=None,
-        feature=context.feature.name,
-        feature_version=misc.define_feature_version(context),
-        severity=OutcomeSeverity.EXECUTED,
-        validation_task_id=context.validation_task_id
-    )
-    context.gherkin_outcomes.append(validation_outcome)
+    #ensure the rule is not activated when there are no instances
+    #in case there are no instances but the rule is applicable (e.g. SPS001), then the rule is still activated and will return either a pass or an error
+    is_activated = any(misc.recursive_flatten(instances)) if instances else context.applicable
+    if is_activated:
+         # in case we're testing, add a custom behave error to flag the rule as passed for the CI/CD pipeline
+        if context.config.userdata.get('execution_mode') and eval(context.config.userdata.get('execution_mode')) == ExecutionMode.TESTING:
+            validation_outcome = ValidationOutcome(
+                    outcome_code=ValidationOutcomeCode.PASSED,
+                    feature=context.feature.name,
+                    observed=None, 
+                    expected=None,
+                    feature_version=misc.define_feature_version(context),
+                    severity= OutcomeSeverity.ERROR,
+                    validation_task_id=context.validation_task_id
+                )
+        else: # in Production, the rule is marked as green when executed and there are no subsequent errors
+            validation_outcome = ValidationOutcome(
+                outcome_code=ValidationOutcomeCode.EXECUTED,  # "Executed", but not no error/pass/warning #deactivated for now
+                observed="passed",
+                expected=None,
+                feature=context.feature.name,
+                feature_version=misc.define_feature_version(context),
+                severity=OutcomeSeverity.EXECUTED,
+                validation_task_id=context.validation_task_id
+            )
+        context.gherkin_outcomes.append(validation_outcome)
```

Comment:
I think I understand what you're trying to do, but it's rather hacky for two reasons:

- the combination of `OutcomeSeverity.ERROR` and `ValidationOutcomeCode.PASSED`
- the difference in behaviour between production and local testing

This is bound to confuse people in the future, us included.

I don't think it needs to be this way. We just need to figure out how we get information from the gherkin_outcomes into the json log.

I've stumbled upon something called `context.attach` that seems to be a way to propagate some information all the way the json log, but I think we could also write our own JsonFormatter subclass that picks information from the outcomes up.

https://github.com/behave/behave/commit/29b6d471c7d73b894753f85fcf161fa7b5d58105#diff-23cf37bbf1511c05d382150305e9f11f0e2a6bfcfcfd7de211d588c1e54f7585R136